### PR TITLE
split gotodefinition tests in input/expected files

### PR DESCRIPTION
**Pull request by smoothdeveloper**
refactor gotodefinition tests (37 refactored over 57, 20 remaining needs another pass of extraction or manual extraction)

#1444

### CODE REVIEW

**Comment by dungpa** [comment]
Looks much better (than current master).

Conceptually the expected files should be signature files (`*.fsi`).

IMO a folder for each test is a bit overkill. We can have a flat structure with three files for each test:

<test_case>.input.fs
<test_case>.expected.fsi
<test_case>.settings.json

It's your choice though.

**Comment by smoothdeveloper** [comment]
The folder (I agree the nesting is abit annoying) simplifies the iteration in the implementation. I'd rather keep it as is for now but can change if someone else also prefer a flat layout.

I should be looking at printing a proper diff at some point also.

**Comment by cloudRoutine** [comment]
I agree with @dungpa, it would be nicer without all the sub-directories

For the iteration just group by substring before the first `.` then use a recursive loop to pattern match by the file extension and put the paths into a record which you use to provide the params to your test function. Easy peasy :wink:

**Comment by smoothdeveloper** [comment]
I'm still considering maintenance pain, and it feels that adding a new test is easier by copying a directory than manually naming three files.

Would it make sense to keep the directories in this regard?

**Comment by cloudRoutine** [comment]
take a list of all the test names and write a lil script to generate the files, you'll be done in a min or two. Having them all in the same directory make sit easier to include them in the fsproject as non compile files so they can be reviewed/searched in VS

**Comment by smoothdeveloper** [comment]
We can make a csproj and a wildcard for content files so the proj files never changes, it works with subfolders too.

**Comment by smoothdeveloper** [comment]
And it takes more than a minute or two to pull out the correct script here :smile: sorry!

**Comment by cloudRoutine** [comment]
don't add the csproj, it's not worth it

File: tests/FSharp.Editing.Tests/GoToDefinitionTests.fs [hunk 1]
```diff
@@ -34,10 +15,32 @@ open System.Xml.Linq
...
 let xmlFileCache = Dictionary()
 
+type GenericTestsFolder = FSharp.Management.FileSystem< const(__SOURCE_DIRECTORY__ + "/../data/gotodef")>
+type GenericTestSettings = JsonProvider<GenericTestsFolder.``generic-sample``.``settings.list.json``, SampleIsList = true>
+
+let genericCasesDirectory = DirectoryInfo(GenericTestsFolder.``generic-cases``.Path)
+
+type CompareStringResult =
+| Equal
+| NotEqual of message: string
```
Comment (vasily-kirichenko) [inline tests/FSharp.Editing.Tests/GoToDefinitionTests.fs]:
```fsharp
type CompareStringResult =
    | Equal
    | NotEqual of message: string
```

File: tests/FSharp.Editing.Tests/GoToDefinitionTests.fs [hunk 2]
```diff
@@ -34,10 +15,32 @@ open System.Xml.Linq
...
+  let expected = expected.Replace("\r","")
+  let actual = actual.Replace("\r","")
+  use messageWriter = new Internal.TextMessageWriter(MaxLineLength = 30000)
+  let equalConstraint = Constraints.EqualConstraint(expected)
+  let result = equalConstraint.ApplyTo(actual)
+  result.WriteMessageTo(messageWriter)
+  if result.IsSuccess then
+    Equal
+  else
+    NotEqual (messageWriter.ToString().Replace(@"\r",@"").Replace(@"\n", Environment.NewLine))
```
Comment (vasily-kirichenko) [inline tests/FSharp.Editing.Tests/GoToDefinitionTests.fs]: Use 4 spaces indents everywhere.

File: tests/FSharp.Editing.Tests/GoToDefinitionTests.fs [hunk 3]
```diff
@@ -34,10 +15,32 @@ open System.Xml.Linq
 open System.Collections.Generic
 open System.IO
 open FSharp.Editing.Features.SignatureGenerator
+open System.Reflection
```
Comment (vasily-kirichenko) [inline tests/FSharp.Editing.Tests/GoToDefinitionTests.fs]: Isn't used.

File: tests/FSharp.Editing.Tests/GoToDefinitionTests.fs [hunk 4]
```diff
@@ -124,56 +128,98 @@ let generateDefinitionFromPos caretPos src =
...
+| AssertEqualFirstLines of input: string * position: pos * expected: string * firstLines: int
+
+[<NoComparison>]
+type GenericTestResult =
+| Passed
+| Inconclusive
+| Failed of exn
+| IncorrectFolderSetup of message: string
+| IncorrectResult of message: string
+with
```
Comment (vasily-kirichenko) [inline tests/FSharp.Editing.Tests/GoToDefinitionTests.fs]: Remove `with`.

File: tests/FSharp.Editing.Tests/GoToDefinitionTests.fs [hunk 5]
```diff
@@ -124,56 +128,98 @@ let generateDefinitionFromPos caretPos src =
...
+    | "eq"           -> AssertEqual(source, caretPos, expected)
+    | "firstLinesEq" -> AssertEqualFirstLines(source, caretPos, expected, settings.FirstLineCount.Value)
+    | _ -> failwithf "assertType '%s' is invalid, expected 'eq' or 'firstLinesEq'" settings.AssertType
+
+  let executeTest directory =
+    let test =
+      try
+        Choice1Of2 <| parseTest directory
+      with
+      | e -> Choice2Of2 (string e)
```
Comment (vasily-kirichenko) [inline tests/FSharp.Editing.Tests/GoToDefinitionTests.fs]: As we started to use Either, let's introduce `Ok` and `Fail` aliases and `Ok` and `Fail` APs for completeness.

File: tests/FSharp.Editing.Tests/GoToDefinitionTests.fs [hunk 6]
```diff
@@ -124,56 +128,98 @@ let generateDefinitionFromPos caretPos src =
...
+        | _ -> failwithf "unexpected to reach this %s" directory.Name
+        
+  let results = 
+      genericCasesDirectory.EnumerateDirectories()
+      |> Seq.map executeTest
+      |> Seq.toArray
+  System.Console.WriteLine("")
+  let inconclusiveCount =
+    results 
+    |> Seq.filter (function | _, Inconclusive -> true | _ -> false)
```
Comment (vasily-kirichenko) [inline tests/FSharp.Editing.Tests/GoToDefinitionTests.fs]:
```fsharp
|> Seq.filter (snd >> (=) Inconclusive)
```

**Comment by smoothdeveloper** [comment]
@dungpa / @cloudRoutine I'll take the idea of no subfolders for later refactorings in other areas, I'd like to focus on more interesting stuff after layground work :smile: 

@vasily-kirichenko I just changed the items you mentioned.

**Comment by cloudRoutine** [comment]
@smoothdeveloper  why is `/tests/data/gotodef/generic-sample` still left in?

Have you tried running all of the directory based tests in parallel?

**Comment by smoothdeveloper** [comment]
@cloudRoutine the `generic-sample` folder contains the sample json file (used for `JsonProvider`).

I haven't tried to make the test run in parallel.

**Pull request merged**